### PR TITLE
[Cpp worker]Support cpp call java actor

### PR DESCRIPTION
**Pull request by WangTaoTheTonic**
## Why are these changes needed?
Support x-lang invocation from cpp to java actor.

TODO: doc

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: cpp/BUILD.bazel [hunk 1]
```diff
@@ -157,6 +157,7 @@ cc_test(
         "counter.so",
         "plus.so",
         "ray_cpp_pkg",
+        "src/ray/test/cluster/ray-test-SNAPSHOT.jar",
```
Comment (SongGuyang) [inline cpp/BUILD.bazel:160]: Can we move the xlang test to a separate test like `cluster_mode_xlang_test`? In some cases, we need to run the `cluster_mode_test ` in a simple environment which doesn't compile the java code or even doesn't install the JDK.

File: cpp/include/ray/api/arguments.h [hunk 2]
```diff
@@ -58,6 +59,16 @@ class Arguments {
 
         PushValueArg(task_args, std::move(dummy_buf), METADATA_STR_RAW);
         PushValueArg(task_args, std::move(buffer), METADATA_STR_XLANG);
+      } else if (lang_type == LangType::JAVA) {
```
Comment (SongGuyang) [inline cpp/include/ray/api/arguments.h:62]: Can we avoid the duplicated code here?

File: cpp/src/ray/runtime/abstract_ray_runtime.cc [hunk 3]
```diff
@@ -199,6 +199,15 @@ std::string AbstractRayRuntime::CallActor(
     func_holder.class_name = typed_descriptor->ClassName();
     invocation_spec = BuildInvocationSpec1(
         TaskType::ACTOR_TASK, func_holder, args, ActorID::FromBinary(actor));
+  } else if (remote_function_holder.lang_type == LangType::JAVA) {
```
Comment (SongGuyang) [inline cpp/src/ray/runtime/abstract_ray_runtime.cc:202]: Can we avoid the duplicated code here?

File: java/test/src/main/resources/test_cross_language_invocation.py [hunk 4]
```diff
@@ -58,9 +58,7 @@ def py_func_call_java_function():
 @ray.remote
 def py_func_call_java_actor(value):
     assert isinstance(value, bytes)
-    c = ray.cross_language.java_actor_class(
-        "io.ray.test.CrossLanguageInvocationTest$TestActor"
-    )
+    c = ray.cross_language.java_actor_class("io.ray.test.TestActor")
```
Comment (SongGuyang) [inline java/test/src/main/resources/test_cross_language_invocation.py:61]: Why do this change in current PR?

File: cpp/src/ray/runtime/abstract_ray_runtime.cc (hunk 3 above)
Comment (WangTaoTheTonic) [reply cpp/src/ray/runtime/abstract_ray_runtime.cc:202]: Only two lines can be re-used, maybe it's better to keep current structure to save one more indent.

File: java/test/src/main/resources/test_cross_language_invocation.py (hunk 4 above)
Comment (WangTaoTheTonic) [reply java/test/src/main/resources/test_cross_language_invocation.py:61]: The `CrossLanguageInvocationTest.java` includes not only test cases but also the definition parts, which is quite large. We can separate them for clearer. In future we will separate java->cpp and java->python part.

File: ci/ci.sh [hunk 5]
```diff
@@ -229,6 +229,12 @@ test_cpp() {
   # C++ worker example need _GLIBCXX_USE_CXX11_ABI flag, but if we put the flag into .bazelrc, the linux ci can't pass.
   # So only set the flag in c++ worker example. More details: https://github.com/ray-project/ray/pull/18273
   echo build --cxxopt="-D_GLIBCXX_USE_CXX11_ABI=0" >> ~/.bazelrc
+  # prepare cpp worker test resources
+  bazel build //java:copy_pom_file
+  cd java
+  mvn package -DskipTests=true
+  cp test/target/ray-test-*-SNAPSHOT.jar ../cpp/src/ray/test/cluster/ray-test-SNAPSHOT.jar
```
Comment (raulchen) [inline ci/ci.sh:236]: Better use the jar file generated by Bazel.

File: cpp/BUILD.bazel (hunk 1 above)
Comment (raulchen) [inline cpp/BUILD.bazel:160]: Use "//java:libio_ray_ray_test.jar“ instead, then we don't have to manually prepare the jar file.

File: cpp/include/ray/api/actor_handle.h [hunk 6]
```diff
@@ -57,6 +57,14 @@ class ActorHandle {
     return {ray::internal::GetRayRuntime().get(), id_, std::move(remote_func_holder)};
   }
 
+  template <typename R>
+  ray::internal::ActorTaskCaller<JavaActorMethod<R>> Task(JavaActorMethod<R> func) {
+    static_assert(IsXlang, "Actor function type is not match with actor class");
```
Comment (raulchen) [inline cpp/include/ray/api/actor_handle.h:62]:
```suggestion
    static_assert(IsXlang, "Actor function type does not match with actor class");
```

File: cpp/BUILD.bazel (hunk 1 above)
Comment (SongGuyang) [reply cpp/BUILD.bazel:160]: cpp->python is different because c++ test depends on python anyway. I suggest move the java test first in current PR.

Comment (WangTaoTheTonic) [reply cpp/BUILD.bazel:160]: Only x-lang tests depends on python.

Comment (SongGuyang) [reply cpp/BUILD.bazel:160]: All the c++ tests depend on python because we should start ray cluster using `ray start --head`.

**Review by raulchen — APPROVED** [review]

**Pull request merged**